File: webapp/src/main/scala/org/linnando/mixemulator/webapp/VirtualMachineService.scala
package org.linnando.mixemulator.webapp

import angulate2.std._
import org.linnando.mixemulator.{DevicesFrontEnd, VirtualMachineFrontEnd}
import org.linnando.mixemulator.vm.VirtualMachineState
import org.linnando.mixemulator.vm.io.data.IOWord
import rxjs.Subject

import scala.collection.immutable.SortedMap
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

@Injectable()
class VirtualMachineService {
  var text: String =
    """START      OUT  HELLO(18)
           JBUS *(18)
           HLT
HELLO      ALF  HELLO
           ALF  , WOR
           ALF  LD
           END START"""
  var mode = "binary"
  var tracking = true

  val stateChange: Subject[Unit] = new Subject()

  private var maybeMachine: Option[VirtualMachineFrontEnd] = None
  private val devicesFrontEnd = new DevicesFrontEnd(
    new BlockAccessFileOpsImpl,
    new LineAccessFileInputOpsImpl,
    new LineAccessFileOutputOpsImpl)

  def assemble(): Future[Unit] = Future {
    val lines = text.split("\n").toVector
    maybeMachine = Some(
      mode match {
        case "binary" =>
          if (tracking) VirtualMachineFrontEnd.createBinaryTracking(lines, devicesFrontEnd)
          else VirtualMachineFrontEnd.createBinaryNonTracking(lines, devicesFrontEnd)
        case "decimal" =>
          if (tracking) VirtualMachineFrontEnd.createDecimalTracking(lines, devicesFrontEnd)
          else VirtualMachineFrontEnd.createDecimalNonTracking(lines, devicesFrontEnd)
      }
    )
  }

  def go(): Future[Unit] = {
    val eventualMachine: Future[VirtualMachineFrontEnd] = {
      mode match {
        case "binary" =>
          if (tracking) VirtualMachineFrontEnd.goBinaryTracking(devicesFrontEnd)
          else VirtualMachineFrontEnd.goBinaryNonTracking(devicesFrontEnd)
        case "decimal" =>
          if (tracking) VirtualMachineFrontEnd.goDecimalTracking(devicesFrontEnd)
          else VirtualMachineFrontEnd.goDecimalNonTracking(devicesFrontEnd)
      }
    }
    eventualMachine map { machine =>
      maybeMachine = Some(machine)
    }
  }

  def switchOffMachine(): Unit = {
    maybeMachine = None
  }

  def isActive: Boolean = maybeMachine.isDefined

  def canMoveForward: Boolean = maybeMachine.exists(_.canMoveForward)

  def runForward(): Future[Unit] = maybeMachine.map(_.runForward().map(stateChange.next)).getOrElse(Future {})

  def stepForward(): Future[Unit] = maybeMachine.map(_.stepForward().map(stateChange.next)).getOrElse(Future {})

  def canMoveBack: Boolean = maybeMachine.exists(_.canMoveBack)

  def runBack(): Unit = maybeMachine.foreach(m => {
    m.runBack()
    stateChange.next(Unit)
  })

  def stepBack(): Unit = maybeMachine.foreach(m => {
    m.stepBack()
    stateChange.next(Unit)
  })

  def machineState: Option[VirtualMachineState] = maybeMachine.map(_.currentState)

  def symbolsLength: Int = maybeMachine.map(_.symbolsLength).getOrElse(0)

  def symbolIndices: Range = maybeMachine.map(_.symbolIndices).getOrElse(0 until 0)

  def programCounterIndex: Option[Int] = maybeMachine.map(_.programCounterIndex)

  def toggleBreakpointAt(index: Int): Unit = maybeMachine.foreach(_.toggleBreakpointAt(index))

  def addressAt(index: Int): Option[Short] = maybeMachine.flatMap(_.addressAt(index))

  def breakpointAt(index: Int): Option[Boolean] = maybeMachine.map(_.breakpointAt(index))

  def cellContent(index: Int): Option[IOWord] = maybeMachine.flatMap(_.cellContent(index))

  def lineNumberAt(index: Int): Option[Int] = maybeMachine.flatMap(_.lineNumberAt(index))

  def lineAt(index: Int): Option[String] = maybeMachine.flatMap(_.lineAt(index))

  def lineIsModifiedAt(index: Int): Option[Boolean] = maybeMachine.map(_.lineIsModifiedAt(index))

  def deviceNumbers: Iterable[Int] = VirtualMachineService.deviceNames.keys

  def deviceName(deviceNum: Int): String = VirtualMachineService.deviceNames(deviceNum)

  def blockDeviceData(deviceNum: Int): Future[IndexedSeq[IOWord]] =
    maybeMachine.map(_.blockDeviceData(deviceNum)).getOrElse(devicesFrontEnd.getBlockDeviceData(deviceNum))

  def saveBlockDevice(deviceNum: Int, data: Array[Byte]): Future[Unit] =
    devicesFrontEnd.saveBlockDevice(deviceNum, data)

  def lineDeviceData(deviceNum: Int): Future[IndexedSeq[String]] =
    maybeMachine.map(_.lineDeviceData(deviceNum)).getOrElse(devicesFrontEnd.getLineDeviceData(deviceNum))

  def saveLineDevice(deviceNum: Int, data: String): Future[Unit] =
    devicesFrontEnd.saveLineDevice(deviceNum, data)
}

object VirtualMachineService {
  private val deviceNames: Map[Int, String] = SortedMap(
    0 -> "Tape Unit 0",
    1 -> "Tape Unit 1",
    2 -> "Tape Unit 2",
    3 -> "Tape Unit 3",
    4 -> "Tape Unit 4",
    5 -> "Tape Unit 5",
    6 -> "Tape Unit 6",
    7 -> "Tape Unit 7",
    8 -> "Disk Unit 0",
    9 -> "Disk Unit 1",
    10 -> "Disk Unit 2",
    11 -> "Disk Unit 3",
    12 -> "Disk Unit 4",
    13 -> "Disk Unit 5",
    14 -> "Disk Unit 6",
    15 -> "Disk Unit 7",
    16 -> "Card Reader",
    17 -> "Card Punch",
    18 -> "Line Printer",
    20 -> "Paper Tape"
  )
}

File: webapp/src/app/virtual-machine.service.ts
import { Injectable } from '@angular/core';
import { DevicesFrontEnd, DevicesFrontEnd$, IOWord, VirtualMachineFrontEnd, VirtualMachineFrontEnd$, VirtualMachineState } from '@mixemulator/lib';
import { Subject } from 'rxjs';
import { BlockAccessFileOpsImpl } from './block-access-file-ops-impl';
import { LineAccessFileInputOpsImpl } from './line-access-file-input-ops-impl';
import { LineAccessFileOutputOpsImpl } from './line-access-file-output-ops-impl';

@Injectable({
  providedIn: 'root'
})
export class VirtualMachineService {
  private static readonly deviceNames: Map<number, string> = new Map([
    [0, 'Tape Unit 0'],
    [1, 'Tape Unit 1'],
    [2, 'Tape Unit 2'],
    [3, 'Tape Unit 3'],
    [4, 'Tape Unit 4'],
    [5, 'Tape Unit 5'],
    [6, 'Tape Unit 6'],
    [7, 'Tape Unit 7'],
    [8, 'Disk Unit 0'],
    [9, 'Disk Unit 1'],
    [10, 'Disk Unit 2'],
    [11, 'Disk Unit 3'],
    [12, 'Disk Unit 4'],
    [13, 'Disk Unit 5'],
    [14, 'Disk Unit 6'],
    [15, 'Disk Unit 7'],
    [16, 'Card Reader'],
    [17, 'Card Punch'],
    [18, 'Line Printer'],
    [20, 'Paper Tape'],
  ]);

  text = `START      OUT  HELLO(18)
           JBUS *(18)
           HLT
HELLO      ALF  HELLO
           ALF  , WOR
           ALF  LD
           END START`;
  mode = 'binary';
  tracking = true;

  readonly stateChange = new Subject<void>();

  private maybeMachine?: VirtualMachineFrontEnd;
  private readonly devicesFrontEnd: DevicesFrontEnd = DevicesFrontEnd$.create(
    new BlockAccessFileOpsImpl(localStorage),
    new LineAccessFileInputOpsImpl(localStorage),
    new LineAccessFileOutputOpsImpl(localStorage));

  assemble(): void {
    const lines = this.text.split('\n');
    switch (this.mode) {
      case 'binary':
        if (this.tracking) {
          this.maybeMachine = VirtualMachineFrontEnd$.createBinaryTracking(lines, this.devicesFrontEnd);
        } else {
          this.maybeMachine = VirtualMachineFrontEnd$.createBinaryNonTracking(lines, this.devicesFrontEnd);
        }
        break;
      case 'decimal':
        if (this.tracking) {
          this.maybeMachine = VirtualMachineFrontEnd$.createDecimalTracking(lines, this.devicesFrontEnd);
        } else {
          this.maybeMachine = VirtualMachineFrontEnd$.createDecimalNonTracking(lines, this.devicesFrontEnd);
        }
        break;
    }
  }

  async go(): Promise<void> {
    switch (this.mode) {
      case 'binary':
        if (this.tracking) {
          this.maybeMachine = await VirtualMachineFrontEnd$.goBinaryTracking(this.devicesFrontEnd);
        } else {
          this.maybeMachine = await VirtualMachineFrontEnd$.goBinaryNonTracking(this.devicesFrontEnd);
        }
        break;
      case 'decimal':
        if (this.tracking) {
          this.maybeMachine = await VirtualMachineFrontEnd$.goDecimalTracking(this.devicesFrontEnd);
        }
        else {
          this.maybeMachine = await VirtualMachineFrontEnd$.goDecimalNonTracking(this.devicesFrontEnd);
        }
        break;
    }
  }

  switchOffMachine(): void {
    this.maybeMachine = undefined;
  }

  get isActive(): boolean {
    return this.maybeMachine != null;
  }

  get canMoveForward(): boolean {
    return this.maybeMachine?.canMoveForward ?? false;
  }

  async runForward(): Promise<void> {
    await this.maybeMachine?.runForward();
    this.stateChange.next();
  }

  async stepForward(): Promise<void> {
    await this.maybeMachine?.stepForward();
    this.stateChange.next();
  }

  get canMoveBack(): boolean {
    return this.maybeMachine?.canMoveBack ?? false;
  }

  runBack(): void {
    if (this.maybeMachine != null) {
      this.maybeMachine.runBack();
      this.stateChange.next();
    }
  }

  stepBack(): void {
    if (this.maybeMachine != null) {
      this.maybeMachine.stepBack();
      this.stateChange.next();
    }
  }

  get machineState(): VirtualMachineState | undefined {
    return this.maybeMachine?.currentState;
  }

  get symbolsLength(): number {
    return this.maybeMachine?.symbolsLength ?? 0;
  }

  get symbolIndices(): number[] {
    return this.maybeMachine?.symbolIndices ?? [];
  }

  get programCounterIndex(): number | undefined {
    return this.maybeMachine?.programCounterIndex;
  }

  toggleBreakpointAt(index: number): void {
    this.maybeMachine?.toggleBreakpointAt(index);
  }

  addressAt(index: number): number | undefined {
    return this.maybeMachine?.addressAt(index);
  }

  breakpointAt(index: number): boolean | undefined {
    return this.maybeMachine?.breakpointAt(index);
  }

  cellContent(index: number): IOWord | undefined {
    return this.maybeMachine?.cellContent(index);
  }

  lineNumberAt(index: number): number | undefined {
    return this.maybeMachine?.lineNumberAt(index);
  }

  lineAt(index: number): string | undefined {
    return this.maybeMachine?.lineAt(index);
  }

  lineIsModifiedAt(index: number): boolean | undefined {
    return this.maybeMachine?.lineIsModifiedAt(index);
  }

  get deviceNumbers(): IterableIterator<number> {
    return VirtualMachineService.deviceNames.keys();
  }

  deviceName(deviceNum: number): string {
    return VirtualMachineService.deviceNames.get(deviceNum) ?? '';
  }

  blockDeviceData(deviceNum: number): Promise<IOWord[]> {
    return this.maybeMachine?.blockDeviceData(deviceNum) ?? this.devicesFrontEnd.getBlockDeviceData(deviceNum);
  }

  saveBlockDevice(deviceNum: number, data: number[]): Promise<void> {
    return this.devicesFrontEnd.saveBlockDevice(deviceNum, data);
  }

  lineDeviceData(deviceNum: number): Promise<string[]> {
    return this.maybeMachine?.lineDeviceData(deviceNum) ?? this.devicesFrontEnd.getLineDeviceData(deviceNum);
  }

  saveLineDevice(deviceNum: number, data: string): Promise<void> {
    return this.devicesFrontEnd.saveLineDevice(deviceNum, data);
  }
}
